crates/matrix-sdk/src/send_queue/mod.rs

```diff
@@ -1284,6 +1286,199 @@ impl QueueStorage {
+                            &last_upload_file_txn.unwrap(),
+                            upload_thumbnail_txn.clone().into(),
+                            created_at,
+                            DependentQueuedRequestKind::UploadFileOrThumbnail {
+                                content_type: thumbnail_content_type.to_string(),
+                                cache_key: thumbnail_media_request.clone(),
+                                related_to: send_event_txn.clone(),
+                                parent_is_thumbnail_upload: false,
+                            },
+                        )
+                        .await?;
+
+                    // Save the file upload request as a dependent request of the thumbnail upload.
+                    store
+                        .save_dependent_queued_request(
+                            &self.room_id,
+                            &upload_thumbnail_txn,
+                            upload_file_txn.clone().into(),
+                            created_at,
+                            DependentQueuedRequestKind::UploadFileOrThumbnail {
+                                content_type: content_type.to_string(),
+                                cache_key: file_media_request.clone(),
+                                related_to: send_event_txn.clone(),
+                                parent_is_thumbnail_upload: true,
+                            },
+                        )
+                        .await?;
+
+                    Some(thumbnail_info)
+                } else {
+                    // Save the file upload as a dependent request of the last file upload.
+                    store
+                        .save_dependent_queued_request(
+                            &self.room_id,
+                            &last_upload_file_txn.unwrap(),
+                            upload_file_txn.clone().into(),
+                            created_at,
+                            DependentQueuedRequestKind::UploadFileOrThumbnail {
+                                content_type: content_type.to_string(),
+                                cache_key: file_media_request.clone(),
+                                related_to: send_event_txn.clone(),
+                                parent_is_thumbnail_upload: false,
+                            },
+                        )
+                        .await?;
+
+                    None
+                };
+
+                finish_item_infos.push(FinishGalleryItemInfo {
+                    file_upload: upload_file_txn.clone(),
+                    thumbnail_info: thumbnail_info.cloned(),
+                });
+
+                last_upload_file_txn = Some(upload_file_txn.clone());
+            }
+        }
+
+        // Push the request for the event itself as a dependent request of the last file
+        // upload.
+        store
+            .save_dependent_queued_request(
+                &self.room_id,
+                &last_upload_file_txn.unwrap(),
```

Comment:
Hey, don't hide an `unwrap()` like this :D I think this is incorrect, if there's no gallery items in the gallery. This should be impossible based on another previous comment I made, but let's make sure of it by putting this code in the `if ... split_first()` above.

Then, we can reduce the entire function's body indent by one level, by having some kind of `let Some((first, rest)) = ...split_first() else { return Ok(()); }`.

Which means that `last_upload_file_txn` will now always be set, something that the compile ought to be able to infer, so we can make it non `Option`al, and start by not setting it initially.